packages/tldraw/src/lib/ui/context/actions.tsx

```diff
@@ -1565,6 +1578,47 @@ export function ActionsProvider({ overrides, children }: ActionsProviderProps) {
 					}
 				},
 			},
+			{
+				id: 'download-original',
+				label: 'action.download-original',
+				readonlyOk: true,
+				onSelect: async (source) => {
+					const selectedShapes = editor.getSelectedShapes()
+					if (selectedShapes.length === 0) return
+
+					const mediaShapes = selectedShapes.filter((s): s is TLImageShape | TLVideoShape =>
+						supportsDownloadingOriginal(s, editor)
+					)
+
+					// We only want to download originals if we only have media shapes selected
+					if (mediaShapes.length !== selectedShapes.length) return
```

Comment:
i don't know about this bit. i worry that, say, there's some images and they're connected by some arrows. the download will fail b/c the connector shapes are stopping the action. wdyt?

Reply:
In my mind, I think it should be less programmatically strict and more on the human-level of thinking.
So the way I would frame it is a graceful degradation of actions around the content that you're interacting with.

For example, on a website, if I copy-paste a page on Wikipedia that includes both text, images, tables. I would expect copy-paste to still work and paste into VSCode, even though I accidentally selected some images which won't be copied into my text editor.
So, I would expect if I selected some shapes which were images that in the context menu the 'Download original' option would be available to click on b/c it's not enforcing a strictness of selection.

Reply:
I think Mime's right, I'd expect it to download everything that can be downloaded. I'd look to other apps here for guidance, it must be a common pattern.

Reply:
Looks like most apps don't even have this feature. They use copy to clipboard or selection export (Figma, FigJam, Whimsical, Excalidraw)

Miro has a download, but only for a single image. Once you select multiple images you don't get that option. Same for Notion.
![CleanShot 2025-06-13 at 12 13 22](https://github.com/user-attachments/assets/aede131b-e0c4-4e17-bb7f-0c2446b24eac)
![CleanShot 2025-06-13 at 12 16 55](https://github.com/user-attachments/assets/5fe5eb7c-ba09-425a-8adc-ccfedd8c5672)

Reply:
I'll loosen the requirement so that this works if at least one downloadable shape is selected 👍